dataflow/encryption-keys/bigquery_kms_key.py

```diff
@@ -0,0 +1,88 @@
+#!/usr/bin/env python
+#
+# Copyright 2019 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+
+
+def run(output_bigquery_table, kms_key, beam_args):
+    # [START dataflow_cmek]
+    import apache_beam as beam
+
+    # output_bigquery_table = '<project>:<dataset>.<table>'
+    # kms_key = 'projects/<project>/locations/<kms-location>/keyRings/<kms-keyring>/cryptoKeys/<kms-key>' # noqa
+    # beam_args = [
+    #     '--project', 'your-project-id',
+    #     '--runner', 'DataflowRunner',
+    #     '--temp_location', 'gs://your-bucket/samples/dataflow/kms/tmp',
+    #     '--region', 'us-central1',
+    # ]
+
+    # Query from the NASA wildfires public dataset:
+    # https://console.cloud.google.com/bigquery?p=bigquery-public-data&d=nasa_wildfire&t=past_week&page=table
+    query = """
+        SELECT latitude,longitude,acq_date,acq_time,bright_ti4,confidence
+        FROM `bigquery-public-data.nasa_wildfire.past_week`
+        LIMIT 10
+    """
+
+    # Schema for the output BigQuery table.
+    schema = {
+        'fields': [
+            {'name': 'latitude', 'type': 'FLOAT'},
+            {'name': 'longitude', 'type': 'FLOAT'},
+            {'name': 'acq_date', 'type': 'DATE'},
+            {'name': 'acq_time', 'type': 'TIME'},
+            {'name': 'bright_ti4', 'type': 'FLOAT'},
+            {'name': 'confidence', 'type': 'STRING'},
+        ],
+    }
+
+    options = beam.options.pipeline_options.PipelineOptions(beam_args)
+    with beam.Pipeline(options=options) as pipeline:
+        (
+            pipeline
+            | 'Read from BigQuery with KMS key' >>
+            beam.io.Read(beam.io.BigQuerySource(
+                query=query,
+                use_standard_sql=True,
+                kms_key=kms_key,
+            ))
+            | 'Write to BigQuery with KMS key' >>
+            beam.io.WriteToBigQuery(
+                output_bigquery_table,
+                schema=schema,
+                write_disposition=beam.io.BigQueryDisposition.WRITE_TRUNCATE,
+                kms_key=kms_key,
```

Comment:
Interesting.  I would have expected some kind of bigquery namespace e.g. beam.io.*, does beam do some cleverness to collapse this or is it just a very big namespace with all the enums for all sources/sinks?

Reply:
Beam does some cleverness in the `__init__.py` files for every directory level. They import anything from certain modules. It is in fact located at `apache_beam.io.gcp.bigquery. WriteToBigQuery`, but it is collapsed (and thus accessible) simply as `apache_beam.io.WriteToBigQuery` :)